File: Zhang_rho/src/get_sky_sun_rho.py
import argparse
from functools import partial
import pathlib
import sys

import numpy as np
import xarray as xr

π = np.pi

db_path = '../db.mat'
#Groups in db.mat
db, quads, sdb, vdb = None, None, None, None
#Vars. in db.mat
skyrad0, sunrad0, rad_boa_sca, rad_boa_vec = None, None, None, None


def parse_command_line(args):
    'Maps command line arguments to available options.'

    parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter, 
                                     description='Generates ToA Radiance')
    parser.add_argument('-i', '--input_path', type=str, required='True', 
                        help='path to l2 file with Lt available')
    parser.add_argument('-t', '--input_type', type=str, default=)
    parser.add_argument('-o', '--output', type=str, help='Output filename', default='Lt')
    parser.add_argument('-l', '--lines', type=int, default=2030, 
                        help='number of lines to process (top-down)')
    parser.add_argument('-p', '--pixels', type=int, default=1354,
                        help='number of pixels to process (left-right)')
    return parser.parse_args(args)


def gen_vec_quad():
    pass


def get_prob(wind, vec,):
    """
    Computes probability of sky light being reflected into the seson

    Inputs
    ------
    wind :
    vec :

    Outputs
    -------
    prob [np.array] :  probability of sky light reflected into the sensor
    angr_sky[np.array] : reflection angle
    """
    
    with xr.open_dataset(db_path, group='quads') as quads:
        prob = np.Nan * np.ones((quads.zen.size, wind.size*vec.shape[0]))
    angr_sky = prob.copy()
    
    angr_sky, prob = skylight_reflection2(wind, vec)    
    
def skylight_reflection2(wind, sensor, quads):
    """
    Computes probability of light reflection at angle.

    Inputs
    ------
    wind :
            Wind speed (m/s)
    sensor : Numpy array
             The vector of reflected light measured by the sensor

    quads :
            Sky light quads
    """
    # ...
    # p_vec = gen_vec_polar(zen0, quads.sun05, num)
    # prob, ang = prob_reflection(-p_vec, sensor, wind)
    # ...
    # half_azm = np.linspace()
    # sky = gen_vec(quads.zen...)
    # prob, ang = prob_reflection(...)
    

def my_sph2cart(azm, zen, r=1):
    """
    Converts spherical coordinates to cartesian

    Inputs
    -------
    azm [Numpy array] : -azimuth angle 
    zen [Numpy array] : zenith angle
    r  [float] : radius = 1

    Outputs
    -------
    x :
    y :
    z :
    """
    def sph2cart(azm, elev, r):
        cos_elev = np.cos(elev)
        x = r * cos_elev * np.cos(azm)
        y = r * cos_elev * np.sin(azm)
        z = r * np.sin(elev)
        return x, y, z
    x, y, z = sph2cart(azm, π/2 - zen, r)
    return np.c_[x.ravel(), y.ravel(), z.ravel()].squeeze()


def find_quads(zen, azm):
    """
    Finds location in quads (why is it called find_quads?)

    Inputs
    ------
    zen :
    azm :

    Outputs
    -------
    locs :
    """
    loc = None
    try:
        with xr.open_dataset(db_path, group='quads') as quads:
            tmp = np.sqrt((quads.zen[:] - zen)**2 + (quads.azm[:] - azm)**2)
            loc = np.argmin(tmp)
    except:
        pass
    finally:
        return loc

def sw_fresnel():
    """
    Calcualtes Fresnel reflectance for seawater.

    Inputs
    ------
    wv :
         Wavelength (nm)
    ang :
          Reflectance angle
    T :
        Temperature (̊ C)
    S :
        Salinity (PSU)

    Outputs
    -------
    m :
        Refractive index
    ref :
         Fresnel reflectance of seawater
    """
    pass

def index_w(wv, T, S):
    """
    Calculates water refractive index

    Inputs
    -------
    wv :
         Wavelength (nm)
    T :
    Temperature (̊ C)
    S :
        Salinity (PPT)
    """
    pass

def fresnel(m ,ang):
    """
    Calculates Fresnel reflectances.

    Inputs
    ------
    m :
        Refractive index
    ang :
        Reflectance angle

    Outputs
    -------
    R :
        Fresnel reflectance matrix element (1, 1)
    R12 :
          Fresnel reflectance matrix element (1, 2)
    R33 :
          Fresnel reflectance matrix element (3, 3)
    """
    pass

def gen_vec_polar(zen, sun05, num=10):
    """
    Generates vectros for the polar cap, quad, and sun disk.
    By convention, the sun disk is at XZ plane, i.e., azimuth = 0.

    Inputs
    ------
    zen :
         Sun zenith angle

    sun05 :

    num :
          Number of angles to consider
    Outputs
    -------
    vec :
          Polar cap vector
    """
    
    ϕ = np.linspace(0, 2*π, 100)
    sin_sun05 = np.sin(sun05)
    tmp = np.c_[sin_sun05*np.cos(ϕ), sin_sun05 * np.sin(ϕ), np.cos(sun05)*np.ones_like(ϕ)]
    tmp = np.insert(tmp, 0, [0, 0, 1], axis=0)
    ry = np.c_[[np.cos(zen), 0, np.sin(zen)], 
               [0, 1, 0], 
               [-np.sin(zen), 0, np.cos(zen)]]
    vec = tmp @ ry
    return vec    

def prob_reflection(inc, refl, wind):
    """
    Estimates probability of facets reflecting incident ligth into given direction and wind.

    Inputs
    ------
    inc :
          incident light vector
    refl :
            reflected light vector
    wind :
            wind speeds

    Outputs
    -------
    prob :
           Probability
    ang :
            Reflection angle
    """
    pass


def get_sky_run_rho(env, sensor):
    """
    Computes sea surface reflectance of skylight.

    Inputs
    ------

    Outputs
    -------
    ρ :
        Sea surface reflectance of skylight
    """
    # calls:
    #   deg2rad()✅
    #   my_sph2cart()✅
    #   deg2rad()✅
    #   find_quads()
    #   get_prob()
    #   sum()
    #   sw_Fresnel()
    #   squeeze(interpn())
    #   bsxfun()
    #   sum(bsxfun())
    #   squeeze(interpn())
    #   get_vec_polar(deg2rad())
    #   prob_reflection()

   
"""
def load_db(db_path='./db.mat'):
    global db, quads, sdb, vdb
    global skyrad0, sunrad0, rad_boa_sca, rad_boa_vec
    
    db = xr.open_dataset(db_path, group='db')
    quads = xr.open_dataset(db_path, group='quads')
    sdb = xr.open_dataset(db_path, group='sdb')
    vdb = xr.open_dataset(db_path, group='vdb')
    skyrad0 = xr.open_dataset(db_path)['skyrad0']
    sunrad0 = xr.open_dataset(db_path)['sunrad0']
    rad_boa_sca = xr.open_dataset(db_path)['Radiance_BOA_sca']
    rad_boa_vec = xr.open_dataset(db_path)['Radiance_BOA_vec']
"""



def main(args):
    env_vars = ['wind', 'od', 'C', 'zen_sun', 'wtem', 'sal']
    pargs = parse_command_line(args)    
    with open(pargs.env_file_path) as fenv:
        rows = fenv.readlines()
        env = {k:v for k,v in zip(rows[0].strip().split(','),
                                  np.array(rows[1].strip().split(','), 
                                           dtype='f2')
                                  )
               }
    sensor = dict.fromkeys(['ang', 'wv', 'ang2'])
    sensor['ang'] = np.array([40, 45])
    sensor['wv'] = np.arange(350, 1010, 10)
    sensor['ang2'] = sensor['ang'] + np.array([0, 180])
    sensor['pol'] = np.deg2rad(sensor['ang'])
    sensor['vec'] = my_sph2cart(sensor['pol'][1], sensor['pol'][0])
    sensor['loc2'] = find_quads(*sensor['pol2'])
    
    # Probability and reflection angle of reflecting skylight into the sensor
    prob, angr_sky = get_prob(env['wind'], sensor['vec'])
                                
                                
"""
if __name__ == "__main__":
    main(sys.argv[1:])
"""

File: Zhang_rho/Sky_Light_Reflection/get_sky_sun_rho.m
function rho = get_sky_sun_rho(env, sensor)
persistent db quads skyrad0 sunrad0 sdb vdb Radiance_BOA_sca Radiance_BOA_vec
if isempty(db) || isempty(quads) || isempty(skyrad0) || isempty(sunrad0) || ...
        isempty(sdb) || isempty(vdb) || isempty(Radiance_BOA_sca) || ...
        isempty(Radiance_BOA_vec)
    load('db.mat');
end
tmp = [1:size(skyrad0,3)];

sensor.pol = deg2rad(sensor.ang); % the sensor polar coordinate
sensor.vec = my_sph2cart(sensor.pol(2),sensor.pol(1),1); % sensor vector
sensor.pol2 = deg2rad(sensor.ang2); % the skylight polar coordinate
sensor.loc2 = find_quads(quads,sensor.pol2(1),sensor.pol2(2));

% Probability and reflection angle of reflecting skylight into the sensor
[prob,angr_sky] = get_prob(env.wind,sensor.vec,quads);
tprob = sum(prob,1);
ref = sw_Fresnel(sensor.wv,angr_sky,env.wtem,env.sal);

skyrad=squeeze(interpn(db.zen_sun,db.od,tmp,db.wv,skyrad0,...
    env.zen_sun,env.od,tmp,sensor.wv));
N0 = skyrad(sensor.loc2,:);
N = bsxfun(@rdivide,skyrad,N0);
rho.sky = sum(bsxfun(@times,ref.*N,prob/tprob),1);

sunrad=squeeze(interpn(db.zen_sun,db.od,db.wv,sunrad0,...
    env.zen_sun,env.od,sensor.wv))'; % make it a row vector
sun_vec=gen_vec_polar(deg2rad(env.zen_sun),quads.sun05);
[prob_sun,angr_sun]=prob_reflection(-sun_vec,sensor.vec,env.wind);
ref_sun = sw_Fresnel(sensor.wv,angr_sun,env.wtem,env.sal);
rho.sun=(sunrad./N0).*(ref_sun*prob_sun/tprob);

rad_inc_sca = squeeze(interpn(...
    sdb.wind,sdb.od(:,10),sdb.zen_sun,sdb.wv,sdb.zen_view,sdb.azm_view, ...
    Radiance_BOA_sca, ...
    env.wind,env.od,env.zen_sun,sensor.wv,180-sensor.ang(1),180-sensor.ang(2)));
rad_mea_sca = squeeze(interpn(...
    sdb.wind,sdb.od(:,10),sdb.zen_sun,sdb.wv,sdb.zen_view,sdb.azm_view, ...
    Radiance_BOA_sca, ...
    env.wind,env.od,env.zen_sun,sensor.wv,sensor.ang(1),180-sensor.ang(2)));
rho_sca = rad_mea_sca./rad_inc_sca;

rad_inc_vec = squeeze(interpn(...
    vdb.wind,vdb.od(:,10),vdb.zen_sun,vdb.wv,vdb.zen_view,vdb.azm_view, ...
    Radiance_BOA_vec, ...
    env.wind,env.od,env.zen_sun,sensor.wv,180-sensor.ang(1),180-sensor.ang(2)));
rad_mea_vec = squeeze(interpn(...
    vdb.wind,vdb.od(:,10),vdb.zen_sun,vdb.wv,vdb.zen_view,vdb.azm_view, ...
    Radiance_BOA_vec, ...
    env.wind,env.od,env.zen_sun,sensor.wv,sensor.ang(1),180-sensor.ang(2)));
rho_vec = rad_mea_vec./rad_inc_vec;
rho.sca2vec = rho_vec./rho_sca;
rho.sca2vec = (rho.sca2vec(:))';
rho.rho = rho.sky.*rho.sca2vec + rho.sun;
end